runtime/altair/src/lib.rs

```diff
@@ -84,6 +87,224 @@ pub fn native_version() -> NativeVersion {
+					deposit,
+				}
+			})
+			.collect();
+
+		<pallet_collator_selection::Candidates<T>>::set(infos);
+
+		consumed + Self::db_access_weights(Some(2), Some(1 + 1 * (candidates.len() as u64)))
+	}
+
+	fn inject_candidacy_bond(bond: BalanceOfCollatorSelection<T>) -> Weight {
+		<pallet_collator_selection::CandidacyBond<T>>::set(bond);
+
+		Self::db_access_weights(None, Some(1))
+	}
+
+	fn db_access_weights(reads: Option<u64>, writes: Option<u64>) -> Weight {
+		let mut weight: Weight = 0u32 as Weight;
+
+		if let Some(num_reads) = reads {
+			weight += T::DbWeight::get()
+				.reads(1 as Weight)
+				.saturating_mul(num_reads);
+		}
+
+		if let Some(num_writes) = writes {
+			weight += T::DbWeight::get()
+				.writes(1 as Weight)
+				.saturating_mul(num_writes);
+		}
+
+		weight
+	}
+
+	fn set_keys(from: &[(T::AccountId, T::Keys)]) -> Weight {
+		let inner_set_keys = |who, keys: T::Keys| {
+			let old_keys = <pallet_session::NextKeys<T>>::get(&who);
+
+			for id in T::Keys::key_ids() {
+				let _key = keys.get_raw(*id);
+
+				// TODO(frederik): Do we need this
+				// ensure keys are without duplication.
+				// let is_owner = <pallet_session::KeyOwner<T>>::get((id, key).map_or(true, |owner| &owner == who);
+			}
+
+			for id in T::Keys::key_ids() {
+				let key = keys.get_raw(*id);
+
+				if let Some(old) = old_keys.as_ref().map(|k| k.get_raw(*id)) {
+					if key == old {
+						continue;
+					}
+
+					<pallet_session::KeyOwner<T>>::remove((*id, old));
+				}
+
+				<pallet_session::KeyOwner<T>>::insert((*id, key), &who);
+			}
+
+			<pallet_session::NextKeys<T>>::insert(&who, &keys);
+		};
+
+		from.iter().for_each(|(who, keys)| {
+			inner_set_keys(
+				// TODO(frederik): think of something to recover from the convert...
+				<T as pallet_session::Config>::ValidatorIdOf::convert(who.clone()).unwrap(),
+				keys.clone(),
+			)
+		});
+
+		Self::db_access_weights(Some(3 * (from.len() as u64)), Some(2 * (from.len() as u64)))
+	}
+
+	#[allow(non_snake_case)]
+	fn into_T_tuple(raw_ids: Vec<[u8; 32]>) -> Vec<(T::AccountId, T::Keys)> {
+		raw_ids
+			.to_vec()
+			.into_iter()
+			.map::<(T::AccountId, T::Keys), _>(|bytes| {
+				(
+					AccountPublic::from(sp_core::sr25519::Public(bytes))
+						.into_account()
+						.into(),
+					SessionKeys {
+						aura: sp_consensus_aura::sr25519::AuthorityId::from(
+							sp_core::sr25519::Public(bytes),
+						),
+					}
+					.into(),
+				)
+			})
+			.collect::<Vec<(T::AccountId, T::Keys)>>()
+	}
+}
+
+impl<T> OnRuntimeUpgrade for IntegrateCollatorSelection<T>
+where
+	T: pallet_session::Config + pallet_collator_selection::Config + frame_system::Config,
+	BalanceOfCollatorSelection<T>: From<u128>,
+	T::AccountId: From<sp_runtime::AccountId32>,
+	T::Keys: From<SessionKeys>,
+	[u8; 32]: From<<T as pallet_session::Config>::ValidatorId>,
+{
+	fn on_runtime_upgrade() -> Weight {
+		let mut consumed: Weight = 0;
+
+		let current_validators_ids: Vec<[u8; 32]> = <pallet_session::Validators<T>>::get()
```

Comment:
@mustermeiszer I had to change this logic to load the collators from the session. The main reason being that in the other approach we wouldn't be able to test out the migration & health of this upgrade in our testnet, since the collator keys are different, and we would have bricked the testnet after session change. 

That being said, do you know if there is any more succinct way of doing this conversion from ValidatorId -> [u8;32] ?